py3status/modules/playerctl.py

```diff
@@ -0,0 +1,358 @@
+    players = []
+    seek_delta = 5
+    thresholds = {
+        "status": [("Playing", "good"), ("Paused", "degraded"), ("Stopped", "bad")]
+    }
+    volume_delta = 10
+
+    class Meta:
+        update_config = {
+            "update_placeholder_format": [
+                {
+                    # Escape the title by default because many contain special characters
+                    "placeholder_formats": {"title": ":escape"},
+                    "format_strings": ["format_player"],
+                }
+            ]
+        }
+
+    def post_config_hook(self):
+        self.color_paused = self.py3.COLOR_PAUSED or self.py3.COLOR_DEGRADED
+        self.color_playing = self.py3.COLOR_PLAYING or self.py3.COLOR_GOOD
+        self.color_stopped = self.py3.COLOR_STOPPED or self.py3.COLOR_BAD
+
+        self.thresholds_init = self.py3.get_color_names_list(self.format_player)
+        self.position = self.py3.format_contains(self.format_player, "position")
+        self.cache_timeout = getattr(self, "cache_timeout", 1)
+
+        # Initialize the manager + event listeners
+        self.manager = Playerctl.PlayerManager()
+        self.manager.connect("name-appeared", self._on_name_appeared)
+        self.manager.connect("name-vanished", self._on_name_vanished)
+
+        # Initialize all currently active players
+        for player_name in self.manager.props.player_names:
+            self._init_player(player_name)
+
+        self._start_loop()
+
+    def _player_should_be_tracked(self, player_name):
+        for _filter in self.players:
+            if fnmatch(player_name.name, _filter):
+                return True
+
+        return len(self.players) == 0
+
+    def _init_player(self, player_name):
+        if not self._player_should_be_tracked(player_name):
+            return
+
+        player = Playerctl.Player.new_from_name(player_name)
+
+        # Initialize player event listeners
+        player.connect("loop-status", self._loop_status_changed, self.manager)
+        player.connect("metadata", self._on_metadata, self.manager)
+        player.connect("playback-status", self._status_changed, self.manager)
+        player.connect("seeked", self._on_seeked, self.manager)
+        player.connect("shuffle", self._shuffle_changed, self.manager)
+        player.connect("volume", self._volume_changed, self.manager)
+
+        self.manager.manage_player(player)
+
+    def _start_loop(self):
+        loop = GLib.MainLoop()
+
+        # The loop is blocking so it needs to be run a separate thread
+        thread = Thread(target=loop.run)
+        thread.daemon = True
+        thread.start()
+
+    def _on_name_appeared(self, manager, player_name):
+        self._init_player(player_name)
+
+    def _on_name_vanished(self, manager, name):
+        # Add a very small delay to give Playerctl time to remove the player
+        time.sleep(0.01)
+        self.py3.update()
+
+    def _loop_status_changed(self, player, loop_status, manager):
```

Comment:
I could point these event handlers (loop, volume, metadata, etc.) to the same function to simplify the code a little. However, that might not be best practice since the parameters are of different types (even though it is the same number of them). Additionally, it wouldn't allow extensibility in the future.